File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.py
import heapq
import sys

max_pq = []
min_pq = []

N = int(sys.stdin.readline())

for i in range(N):
    num = int(sys.stdin.readline())
    min_size = len(min_pq)
    max_size = len(max_pq)
    if min_size == max_size:
        if min_size != 0 and min_pq[0] < num:
            heapq.heappush(max_pq, (-min_pq[0], min_pq[0]))
            heapq.heappop(min_pq)
            heapq.heappush(min_pq, num)
        else:
            heapq.heappush(max_pq, (-num, num))
    elif max_size > min_size:
        if max_pq[0][1] > num:
            heapq.heappush(min_pq, max_pq[0][1])
            heapq.heappop(max_pq)
            heapq.heappush(max_pq, (-num, num))
        else:
            heapq.heappush(min_pq, num)
    print(max_pq[0][1])

File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.cc
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    priority_queue<int> max_pq;
    priority_queue<int, vector<int>, greater<int>> min_pq;

    int maxSize = 0;
    int minSize = 0;
    int N, num;
    cin >> N;

    while(N--) { // 7
        cin >> num;
        if (maxSize == minSize) {
            if (minSize != 0 && min_pq.top() < num) {
                max_pq.push(min_pq.top());
                min_pq.pop();
                min_pq.push(num);
                maxSize++;
            } else {
                max_pq.push(num);
                maxSize++;
            }
        } else if (maxSize > minSize) {
            if (max_pq.top() > num) {
                min_pq.push(max_pq.top());
                max_pq.pop();
                max_pq.push(num);
                minSize++;
            } else {
                min_pq.push(num);
                minSize++;
            }
        }

        cout << max_pq.top() << '\n';
    }


    return 0;
}